File: frontend/src/types/department.ts
/**
 * Department 도메인 wire 타입 (A16, docs/02 §7.x, ADR #36).
 *
 * backend `DepartmentSummaryDto` record (`com.ibizdrive.department.DepartmentSummaryDto`)와 wire 1:1.
 * `GET /api/departments/search` 응답의 `items[*]` 형상.
 *
 * `path`(LTREE)/`createdAt` 등은 share subject picker UX에 불필요 → backend가 의도적으로 미노출
 * (ADR #36: minimal surface — A14 UserSummary 답습 + LTREE는 v1.x 트리 쿼리 도입 시 별도 wire).
 */
export interface DepartmentSummary {
  id: string
  name: string
}

/**
 * Admin 부서 목록/생성/수정 응답 항목 (admin-department-crud, Wave 2 T4, docs/02 §7.x).
 *
 * <p>backend {@code AdminDepartmentSummaryResponse} record와 wire 1:1. share-picker용
 * {@link DepartmentSummary}와 분리한 이유:
 * <ul>
 *   <li>share picker는 활성 부서만 노출 + minimal surface (ADR #36)</li>
 *   <li>admin은 비활성 포함 + {@code isActive}/{@code createdAt} 필요 (lifecycle UX)</li>
 * </ul>
 *
 * <p>{@code isActive}는 backend가 {@code deletedAt == null}로 도출한 boolean — UI는 본 필드만 사용.
 */
export interface AdminDepartmentSummary {
  id: string
  name: string
  isActive: boolean
  /** ISO-8601 (OffsetDateTime). */
  createdAt: string
}

/**
 * Spring {@code Page<AdminDepartmentSummaryResponse>} 직렬화 모양 — admin-department-crud.
 *
 * <p>{@link import('./user').AdminUserPage}와 동형. 전체 필드 중 UI에 필요한 부분집합만 type-narrow.
 */
export interface AdminDepartmentPage {
  content: AdminDepartmentSummary[]
  totalElements: number
  totalPages: number
  number: number
  size: number
}

/**
 * Admin 부서 생성 body — backend {@code AdminDepartmentCreateRequest} 1:1 mirror.
 *
 * <p>trim+검증은 backend에서 다시 수행 — UI는 입력값만 그대로 송신.
 */
export interface AdminDepartmentCreateBody {
  name: string
}

/**
 * Admin 부서 PATCH body — backend {@code AdminDepartmentPatchRequest} 1:1 mirror.
 *
 * <p>두 필드 모두 optional. 둘 다 비면 backend 400 ({@code AdminBadPatchException}).
 * {@code isActive=false} → deactivate, {@code true} → reactivate, {@code name} → rename.
 */
export interface AdminDepartmentPatchBody {
  name?: string
  isActive?: boolean
}

File: backend/src/main/java/com/ibizdrive/department/Department.java
package com.ibizdrive.department;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

import java.time.OffsetDateTime;
import java.util.UUID;

/**
 * 부서 엔티티 — Flyway V7 schema에 mapping (A16, ADR #36, docs/02 §2.x).
 *
 * <p>DB schema가 진실의 출처 (CLAUDE.md §3 원칙 6). {@code spring.jpa.hibernate.ddl-auto: validate}로
 * 컬럼 불일치 시 부팅 실패. 단, application 미사용 컬럼은 entity에서 생략 가능 (validate는 entity의 컬럼이
 * DB에 존재하는지만 검증, 역방향은 무시).
 *
 * <p><b>의도적 생략 (KISS, ADR #36)</b>:
 * <ul>
 *   <li>{@code parent_id} — 조직도 트리는 v1.x 미사용 (flat list 가정).</li>
 *   <li>{@code path} (LTREE) — 트리 쿼리 미사용. schema는 v1.x re-migration 회피용으로 보유.</li>
 * </ul>
 *
 * <p>application은 직속 dept matching만 처리 — A16의 share resolution은 user.department_id 직접 비교
 * (PermissionRepository.findEffective JPQL JOIN).
 */
@Entity
@Table(name = "departments")
public class Department {

    @Id
    @Column(name = "id", nullable = false, updatable = false)
    private UUID id;

    @Column(name = "name", nullable = false, length = 100)
    private String name;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "deleted_at")
    private OffsetDateTime deletedAt;

    protected Department() {
        // JPA
    }

    public Department(UUID id, String name, OffsetDateTime createdAt) {
        this.id = id;
        this.name = name;
        this.createdAt = createdAt;
    }

    public UUID getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }

    public OffsetDateTime getDeletedAt() {
        return deletedAt;
    }

    public boolean isDeleted() {
        return deletedAt != null;
    }

    /**
     * admin-department-crud (Wave 2 T4) — 활성 여부 도출.
     *
     * <p>V7 schema의 `departments` 테이블은 `is_active` 컬럼을 갖지 않으며 {@code deleted_at}만 존재한다.
     * 부서 도메인에서는 "비활성화" = "soft-delete"로 통합한다 (KISS, dept는 user처럼 인증/세션 lifecycle이
     * 없어 별도 boolean이 필요 없음). admin이 비활성화한 부서는 share-picker / admin list / dept resolution에서
     * 모두 동등하게 제외된다.
     */
    public boolean isActive() {
        return deletedAt == null;
    }

    /**
     * admin-department-crud (Wave 2 T4) — 부서 이름 변경.
     *
     * <p>호출자(서비스)는 충돌(`UNIQUE INDEX idx_departments_name_active` — V9) 검출과 audit emit
     * 책임을 진다. 도메인 메서드는 입력 정규화(trim) + 길이 검증(1~100자, V7 컬럼 length=100)만 담당.
     *
     * @throws IllegalArgumentException newName이 null/blank 또는 trim 후 1~100자 범위 밖
     */
    public void rename(String newName) {
        if (newName == null) {
            throw new IllegalArgumentException("name must not be null");
        }
        String trimmed = newName.trim();
        if (trimmed.isEmpty()) {
            throw new IllegalArgumentException("name must not be blank");
        }
        if (trimmed.length() > 100) {
            throw new IllegalArgumentException("name must be at most 100 characters");
        }
        this.name = trimmed;
    }

    /**
     * admin-department-crud (Wave 2 T4) — 부서 비활성화 (soft-delete).
     *
     * <p>이미 비활성이면 idempotent — `deletedAt`을 갱신하지 않고 최초 비활성 시각을 보존한다.
     */
    public void deactivate() {
        if (this.deletedAt == null) {
            this.deletedAt = OffsetDateTime.now();
        }
    }

    /**
     * admin-department-crud (Wave 2 T4) — 부서 재활성화.
     *
     * <p>이미 활성이면 idempotent — no-op.
     */
    public void reactivate() {
        this.deletedAt = null;
    }
}
